tests/utils/test_dates.py

```diff
@@ -29,8 +29,10 @@
 
 class TestDates(unittest.TestCase):
```

Comment:
i think these tests should pass

```python
class TestDates:
    @pytest.mark.parametrize('timezone', ['UTC', 'America/Los_Angeles'])
    def test_days_ago(self, timezone):
        from airflow import settings

        settings.TIMEZONE = pendulum.tz.timezone(timezone)
        from airflow.utils import dates

        today_midnight = pendulum.today(settings.TIMEZONE)
        assert today_midnight.tzinfo == dates.days_ago(2).tzinfo == pendulum.tz.timezone(timezone)
        assert dates.days_ago(0) == today_midnight
        assert dates.days_ago(100) == today_midnight.add(days=-100)
        assert dates.days_ago(0, hour=3) == today_midnight + timedelta(hours=3)
        assert dates.days_ago(0, minute=3) == today_midnight + timedelta(minutes=3)
        assert dates.days_ago(0, second=3) == today_midnight + timedelta(seconds=3)
        assert dates.days_ago(0, microsecond=3) == today_midnight + timedelta(microseconds=3)
```

you will need to add more logic to `days_ago` though.

additionally, i think rather than using moving time `pendulum.today` we should actually use  a fixed time such  as  2021-12-28 so that way we  ensure that days_ago(100)  actually does cross a DST boundary.   i would also add a comment inline above the days_ago(100) assert that explains why we need it (namely  to   test behavior across DST  boundary)

Reply:
Because dates.days_ago uses TIMEZONE in airflow.utils.timezone by `timezone.make_aware()`, I think patching `timezone.TIMEZONE`is more appropriate than `settings.TIMEZONE`. airflow.utils.timezone imports TIMEZONE by
`from airflow.settings import TIMEZONE`

But after patching `timezone.TIMEZONE` by direct assignment, test_round_time() is failed because of timezone difference by remaining patch. This is resolved by patching `timezone.TIMEZONE` by `with mock.patch()`. I'd not talk in detail about this problem. If you wonder, just let me know.

But, There is problem in using utcnow() in days_ago. e.g. In 12/30 00:00 am in UTC, Los Angeles time is 12/29 16:00 pm. `dates.days_ago()` is expected to return 12/29 but it returns 12/30, although timezone is right (Los_Angeles).

`timezone.make_aware()` does not seem to convert utcnow() time in `America/Los_Angeles` appropriately. Is this DST boundary problem you talked about?

Anyway, I'm working on this problem right now. I'll comment as soon as this problem resolved.

For now, test code will be

```python
from unittest import mock

class TestDates:
    @pytest.mark.parametrize('time_zone', ['UTC', 'America/Los_Angeles'])
    def test_days_ago(self, time_zone):
        with mock.patch('airflow.utils.timezone.TIMEZONE', pendulum.tz.timezone(time_zone)):
            today_midnight = pendulum.today(time_zone)

            assert today_midnight.tzinfo == dates.days_ago(2).tzinfo == pendulum.tz.timezone(time_zone)
            assert dates.days_ago(0) == today_midnight
            assert dates.days_ago(100) == today_midnight.add(days=-100) # For the DST boundary test.
            assert dates.days_ago(0, hour=3) == today_midnight + timedelta(hours=3)
            assert dates.days_ago(0, minute=3) == today_midnight + timedelta(minutes=3)
            assert dates.days_ago(0, second=3) == today_midnight + timedelta(seconds=3)
            assert dates.days_ago(0, microsecond=3) == today_midnight + timedelta(microseconds=3)

```

Reply:
Instead of using `datetime.utcnow()`, I just replace it with `datetime.now(timezone.TIMEZONE)`
So dates.days_ago code will be like below.

```python
def days_ago(n, hour=0, minute=0, second=0, microsecond=0):
    """
    Get a datetime object representing `n` days ago. By default the time is
    set to midnight.
    """
    today = datetime.now(timezone.TIMEZONE).replace(
        hour=hour, minute=minute, second=second, microsecond=microsecond
    )
    return today - timedelta(days=n)
```

For DST offset problem, I tested it with several condition and it looks fine with timedelta substraction.

test code:
```python
from datetime import datetime, timedelta
import pendulum

after_dst_end = "2021-11-08"
dst_timezone = pendulum.tz.timezone("America/Los_Angeles")

dt = datetime.fromisoformat(after_dst_end).replace(tzinfo=dst_timezone)
print(dt)
print(dt - timedelta(days=1))
print(dt - timedelta(days=2))
print(dt - timedelta(days=260))
```

result:
```
2021-11-08 00:00:00-08:00 # dt
2021-11-07 00:00:00-07:00 # dt - 1
2021-11-06 00:00:00-07:00 # dt - 2
2021-02-21 00:00:00-08:00 # dt - 260
```